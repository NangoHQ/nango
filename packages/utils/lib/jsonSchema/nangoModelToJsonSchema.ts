import type { NangoModel, NangoModelField } from '@nangohq/types';
import type { JSONSchema7 } from 'json-schema';

/**
 * Converts a list of Nango models to a JSON Schema with all the schemas stored in the definitions property.
 */
export function nangoModelsToJsonSchema(models: NangoModel[]): JSONSchema7 {
    const definitions: Record<string, JSONSchema7> = {};

    for (const model of models) {
        definitions[model.name] = nangoModelToJsonSchema(model);
    }

    return { definitions };
}

function nangoModelToJsonSchema(model: NangoModel): JSONSchema7 {
    const properties: Record<string, JSONSchema7> = {};
    const required: string[] = [];
    let dynamicField: NangoModelField | null = null;

    for (const field of model.fields || []) {
        if (field.dynamic && field.name === '__string') {
            dynamicField = field;
            continue;
        }

        const fieldSchema = nangoFieldToJsonSchema(field);
        properties[field.name] = fieldSchema;

        if (!field.optional) {
            required.push(field.name);
        }
    }

    return {
        type: 'object',
        ...(Object.keys(properties).length > 0 && { properties }),
        ...(required.length > 0 && { required }),
        ...(dynamicField && { additionalProperties: nangoFieldToJsonSchema(dynamicField) })
    };
}

function nangoFieldToJsonSchema(field: NangoModelField): JSONSchema7 {
    if (field.array) {
        return {
            type: 'array',
            items: nangoFieldToJsonSchema({ ...field, array: false })
        };
    }

    if (field.model) {
        if (typeof field.value !== 'string') {
            throw new Error('field is model but value is not a string');
        }

        const modelName = field.value;
        return {
            $ref: `#/definitions/${modelName}`
        };
    }

    if (field.union) {
        if (!Array.isArray(field.value)) {
            throw new Error('field is union but value is not an array');
        }

        return {
            oneOf: field.value.map((v) => nangoFieldToJsonSchema(v))
        };
    }

    if (Array.isArray(field.value)) {
        const properties: Record<string, JSONSchema7> = {};
        const required: string[] = [];
        let dynamicField: NangoModelField | null = null;

        for (const subField of field.value) {
            // It's an array of this field type
            if (subField.name === '0') {
                return nangoFieldToJsonSchema(subField);
            }

            if (subField.dynamic && subField.name === '__string') {
                dynamicField = subField;
                continue;
            }

            properties[subField.name] = nangoFieldToJsonSchema(subField);
            if (!subField.optional) {
                required.push(subField.name);
            }
        }

        return {
            type: 'object',
            ...(Object.keys(properties).length > 0 && { properties }),
            ...(required.length > 0 && { required }),
            ...(dynamicField && { additionalProperties: nangoFieldToJsonSchema(dynamicField) })
        };
    }

    if (field.value === 'any[]') {
        return {
            type: 'array',
            items: {} // Matching CLI behavior for this...
        };
    }

    if (field.value === null) {
        return { type: 'null' };
    }

    if (field.tsType && typeof field.value === 'string' && tsTypeMap[field.value]) {
        return tsTypeMap[field.value] as JSONSchema7;
    }

    if (typeof field.value === 'string') {
        return {
            type: 'string',
            const: field.value
        };
    }

    if (typeof field.value === 'boolean') {
        return {
            type: 'boolean',
            const: field.value
        };
    }

    if (typeof field.value === 'number') {
        return {
            type: 'number',
            const: field.value
        };
    }

    // Fallback to Record<string, string>
    return {
        type: 'object',
        additionalProperties: {
            type: 'string'
        }
    };
}

const tsTypeMap: Record<string, JSONSchema7> = {
    number: { type: 'number' },
    bigint: { type: 'number' },
    boolean: { type: 'boolean' },
    string: { type: 'string' },
    date: { type: 'string', format: 'date-time' },
    undefined: { type: 'null' }, // There's no json schema mapping for this.
    'Record<string, any>': { type: 'object' }
};
