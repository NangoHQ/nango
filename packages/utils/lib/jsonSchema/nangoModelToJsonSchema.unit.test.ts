import { describe, expect, it } from 'vitest';

import { nangoModelsToJsonSchema } from './nangoModelToJsonSchema.js';

import type { NangoModel } from '@nangohq/types';

describe('nangoModelsToJsonSchema', () => {
    it('should handle all primitive types and basic features', () => {
        const models: NangoModel[] = [
            {
                name: 'AllPrimitiveTypes',
                fields: [
                    { name: 'stringField', value: 'string', optional: false },
                    { name: 'numberField', value: 'number', optional: false },
                    { name: 'booleanField', value: 'boolean', optional: false },
                    { name: 'dateField', value: 'date', optional: false },
                    { name: 'optionalString', value: 'string', optional: true },
                    { name: 'unknownType', value: 'uuid', optional: false }, // Should default to string record
                    { name: 'stringArray', value: 'string', array: true, optional: false },
                    { name: 'numberArray', value: 'number', array: true, optional: true }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle nested model references', () => {
        const models: NangoModel[] = [
            {
                name: 'Address',
                fields: [
                    { name: 'street', value: 'string', optional: false },
                    { name: 'city', value: 'string', optional: false },
                    { name: 'zipCode', value: 'string', optional: true }
                ]
            },
            {
                name: 'Country',
                fields: [
                    { name: 'code', value: 'string', optional: false },
                    { name: 'name', value: 'string', optional: false }
                ]
            },
            {
                name: 'DetailedAddress',
                fields: [
                    { name: 'address', value: 'Address', model: true, optional: false },
                    { name: 'country', value: 'Country', model: true, optional: false }
                ]
            },
            {
                name: 'User',
                fields: [
                    { name: 'name', value: 'string', optional: false },
                    { name: 'primaryAddress', value: 'DetailedAddress', model: true, optional: false },
                    { name: 'addresses', value: 'Address', model: true, array: true, optional: true }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle circular model references', () => {
        const models: NangoModel[] = [
            {
                name: 'Category',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    { name: 'name', value: 'string', optional: false },
                    { name: 'parentCategory', value: 'Category', model: true, optional: true },
                    { name: 'subcategories', value: 'Category', model: true, array: true, optional: true }
                ]
            },
            {
                name: 'Post',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    { name: 'title', value: 'string', optional: false },
                    { name: 'author', value: 'User', model: true, optional: false },
                    { name: 'comments', value: 'Comment', model: true, array: true, optional: true }
                ]
            },
            {
                name: 'Comment',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    { name: 'content', value: 'string', optional: false },
                    { name: 'author', value: 'User', model: true, optional: false },
                    { name: 'post', value: 'Post', model: true, optional: false },
                    { name: 'replies', value: 'Comment', model: true, array: true, optional: true }
                ]
            },
            {
                name: 'User',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    { name: 'username', value: 'string', optional: false },
                    { name: 'posts', value: 'Post', model: true, array: true, optional: true },
                    { name: 'comments', value: 'Comment', model: true, array: true, optional: true }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle unions with primitives and models', () => {
        const models: NangoModel[] = [
            {
                name: 'Person',
                fields: [
                    { name: 'name', value: 'string', optional: false },
                    { name: 'age', value: 'number', optional: false }
                ]
            },
            {
                name: 'Company',
                fields: [
                    { name: 'companyName', value: 'string', optional: false },
                    { name: 'employees', value: 'number', optional: false }
                ]
            },
            {
                name: 'Tag',
                fields: [
                    { name: 'label', value: 'string', optional: false },
                    { name: 'color', value: 'string', optional: true }
                ]
            },
            {
                name: 'FlexibleEntity',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    // Union of primitive types
                    {
                        name: 'primitiveUnion',
                        union: true,
                        value: [
                            { name: 'string_option', value: 'string' },
                            { name: 'number_option', value: 'number' },
                            { name: 'boolean_option', value: 'boolean' }
                        ],
                        optional: false
                    },
                    // Union of model references
                    {
                        name: 'entityUnion',
                        union: true,
                        value: [
                            { name: 'person_option', value: 'Person', model: true },
                            { name: 'company_option', value: 'Company', model: true }
                        ],
                        optional: false
                    },
                    // Union with arrays
                    {
                        name: 'mixedUnion',
                        union: true,
                        value: [
                            { name: 'string_option', value: 'string' },
                            { name: 'tags_array_option', value: 'Tag', model: true, array: true }
                        ],
                        optional: true
                    }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle complex arrays with nested structures', () => {
        const models: NangoModel[] = [
            {
                name: 'Metadata',
                fields: [
                    { name: 'key', value: 'string', optional: false },
                    { name: 'value', value: 'string', optional: false }
                ]
            },
            {
                name: 'ArrayShowcase',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    // Array of primitives
                    { name: 'tags', value: 'string', array: true, optional: false },
                    { name: 'scores', value: 'number', array: true, optional: true },
                    // Array of models
                    { name: 'metadata', value: 'Metadata', model: true, array: true, optional: false },
                    // Array with union types
                    {
                        name: 'flexibleArray',
                        array: true,
                        union: true,
                        value: [
                            { name: 'string_option', value: 'string' },
                            { name: 'metadata_option', value: 'Metadata', model: true }
                        ],
                        optional: true
                    }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle empty models array', () => {
        const models: NangoModel[] = [];
        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle model with no fields', () => {
        const models: NangoModel[] = [
            {
                name: 'EmptyModel',
                fields: []
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });

    it('should handle nested objects', () => {
        const models: NangoModel[] = [
            {
                name: 'LinearTeamBase',
                fields: [
                    { name: 'id', value: 'string', optional: false },
                    { name: 'name', value: 'string', optional: false }
                ]
            },
            {
                name: 'TeamsPaginatedResponse',
                fields: [
                    {
                        name: 'teams',
                        array: true,
                        model: true,
                        value: 'LinearTeamBase',
                        optional: false
                    },
                    {
                        name: 'pageInfo',
                        optional: false,
                        value: [
                            {
                                name: 'hasNextPage',
                                array: false,
                                value: 'boolean',
                                optional: false
                            },
                            {
                                name: 'endCursor',
                                union: true,
                                value: [
                                    {
                                        name: '0',
                                        array: false,
                                        value: 'string',
                                        optional: false
                                    },
                                    {
                                        name: '1',
                                        array: false,
                                        value: null,
                                        optional: false
                                    }
                                ],
                                optional: false
                            }
                        ]
                    }
                ]
            }
        ];

        const result = nangoModelsToJsonSchema(models);
        expect(result).toMatchSnapshot();
    });
});
