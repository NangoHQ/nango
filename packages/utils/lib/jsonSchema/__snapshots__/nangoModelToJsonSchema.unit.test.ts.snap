// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nangoModelsToJsonSchema > should handle all primitive types and basic features 1`] = `
{
  "definitions": {
    "AllPrimitiveTypes": {
      "properties": {
        "booleanField": {
          "type": "boolean",
        },
        "dateField": {
          "format": "date-time",
          "type": "string",
        },
        "numberArray": {
          "items": {
            "type": "number",
          },
          "type": "array",
        },
        "numberField": {
          "type": "number",
        },
        "optionalString": {
          "type": "string",
        },
        "stringArray": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "stringField": {
          "type": "string",
        },
        "unknownType": {
          "const": "uuid",
          "type": "string",
        },
      },
      "required": [
        "stringField",
        "numberField",
        "booleanField",
        "dateField",
        "unknownType",
        "stringArray",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle arrays of nested models 1`] = `
{
  "definitions": {
    "ModelWithNestedModelArray": {
      "properties": {
        "nestedModels": {
          "items": {
            "properties": {
              "firstName": {
                "type": "string",
              },
              "lastName": {
                "type": "string",
              },
            },
            "required": [
              "firstName",
              "lastName",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "required": [
        "nestedModels",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle circular model references 1`] = `
{
  "definitions": {
    "Category": {
      "properties": {
        "id": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "parentCategory": {
          "$ref": "#/definitions/Category",
        },
        "subcategories": {
          "items": {
            "$ref": "#/definitions/Category",
          },
          "type": "array",
        },
      },
      "required": [
        "id",
        "name",
      ],
      "type": "object",
    },
    "Comment": {
      "properties": {
        "author": {
          "$ref": "#/definitions/User",
        },
        "content": {
          "type": "string",
        },
        "id": {
          "type": "string",
        },
        "post": {
          "$ref": "#/definitions/Post",
        },
        "replies": {
          "items": {
            "$ref": "#/definitions/Comment",
          },
          "type": "array",
        },
      },
      "required": [
        "id",
        "content",
        "author",
        "post",
      ],
      "type": "object",
    },
    "Post": {
      "properties": {
        "author": {
          "$ref": "#/definitions/User",
        },
        "comments": {
          "items": {
            "$ref": "#/definitions/Comment",
          },
          "type": "array",
        },
        "id": {
          "type": "string",
        },
        "title": {
          "type": "string",
        },
      },
      "required": [
        "id",
        "title",
        "author",
      ],
      "type": "object",
    },
    "User": {
      "properties": {
        "comments": {
          "items": {
            "$ref": "#/definitions/Comment",
          },
          "type": "array",
        },
        "id": {
          "type": "string",
        },
        "posts": {
          "items": {
            "$ref": "#/definitions/Post",
          },
          "type": "array",
        },
        "username": {
          "type": "string",
        },
      },
      "required": [
        "id",
        "username",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle complex arrays with nested structures 1`] = `
{
  "definitions": {
    "ArrayShowcase": {
      "properties": {
        "flexibleArray": {
          "items": {
            "oneOf": [
              {
                "type": "string",
              },
              {
                "$ref": "#/definitions/Metadata",
              },
            ],
          },
          "type": "array",
        },
        "id": {
          "type": "string",
        },
        "metadata": {
          "items": {
            "$ref": "#/definitions/Metadata",
          },
          "type": "array",
        },
        "scores": {
          "items": {
            "type": "number",
          },
          "type": "array",
        },
        "tags": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": [
        "id",
        "tags",
        "metadata",
      ],
      "type": "object",
    },
    "Metadata": {
      "properties": {
        "key": {
          "type": "string",
        },
        "value": {
          "type": "string",
        },
      },
      "required": [
        "key",
        "value",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle dynamic fields 1`] = `
{
  "definitions": {
    "DynamicFields": {
      "properties": {
        "fields": {
          "items": {
            "$ref": "#/definitions/Field",
          },
          "type": "array",
        },
      },
      "required": [
        "fields",
      ],
      "type": "object",
    },
    "Field": {
      "properties": {
        "__string": {
          "oneOf": [
            {
              "$ref": "#/definitions/Field",
            },
            {
              "type": "string",
            },
          ],
        },
      },
      "required": [
        "__string",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle empty models array 1`] = `
{
  "definitions": {},
}
`;

exports[`nangoModelsToJsonSchema > should handle literals 1`] = `
{
  "definitions": {
    "Literals": {
      "properties": {
        "lie": {
          "const": false,
          "type": "boolean",
        },
        "numberLiteral": {
          "const": 1,
          "type": "number",
        },
        "numberLiteralUnion": {
          "oneOf": [
            {
              "const": 1,
              "type": "number",
            },
            {
              "const": 2,
              "type": "number",
            },
            {
              "const": 3,
              "type": "number",
            },
          ],
        },
        "stringLiteral": {
          "const": "nango",
          "type": "string",
        },
        "stringLiteralUnion": {
          "oneOf": [
            {
              "const": "nango",
              "type": "string",
            },
            {
              "const": "mongo",
              "type": "string",
            },
          ],
        },
        "truth": {
          "const": true,
          "type": "boolean",
        },
      },
      "required": [
        "stringLiteral",
        "stringLiteralUnion",
        "truth",
        "lie",
        "numberLiteral",
        "numberLiteralUnion",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle model with no fields 1`] = `
{
  "definitions": {
    "EmptyModel": {
      "properties": {},
      "required": [],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle nested model references 1`] = `
{
  "definitions": {
    "Address": {
      "properties": {
        "city": {
          "type": "string",
        },
        "street": {
          "type": "string",
        },
        "zipCode": {
          "type": "string",
        },
      },
      "required": [
        "street",
        "city",
      ],
      "type": "object",
    },
    "Country": {
      "properties": {
        "code": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "code",
        "name",
      ],
      "type": "object",
    },
    "DetailedAddress": {
      "properties": {
        "address": {
          "$ref": "#/definitions/Address",
        },
        "country": {
          "$ref": "#/definitions/Country",
        },
      },
      "required": [
        "address",
        "country",
      ],
      "type": "object",
    },
    "User": {
      "properties": {
        "addresses": {
          "items": {
            "$ref": "#/definitions/Address",
          },
          "type": "array",
        },
        "name": {
          "type": "string",
        },
        "primaryAddress": {
          "$ref": "#/definitions/DetailedAddress",
        },
      },
      "required": [
        "name",
        "primaryAddress",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle nested objects 1`] = `
{
  "definitions": {
    "LinearTeamBase": {
      "properties": {
        "id": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "id",
        "name",
      ],
      "type": "object",
    },
    "TeamsPaginatedResponse": {
      "properties": {
        "pageInfo": {
          "properties": {
            "endCursor": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "null",
                },
              ],
            },
            "hasNextPage": {
              "type": "boolean",
            },
          },
          "required": [
            "hasNextPage",
            "endCursor",
          ],
          "type": "object",
        },
        "teams": {
          "items": {
            "$ref": "#/definitions/LinearTeamBase",
          },
          "type": "array",
        },
      },
      "required": [
        "teams",
        "pageInfo",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle special cases 1`] = `
{
  "definitions": {
    "SpecialCases": {
      "properties": {
        "array": {
          "items": {},
          "type": "array",
        },
        "bigint": {
          "type": "number",
        },
        "nullValue": {
          "type": "null",
        },
        "undefinedValue": {
          "type": "null",
        },
      },
      "required": [
        "nullValue",
        "undefinedValue",
        "bigint",
        "array",
      ],
      "type": "object",
    },
  },
}
`;

exports[`nangoModelsToJsonSchema > should handle unions with primitives and models 1`] = `
{
  "definitions": {
    "Company": {
      "properties": {
        "companyName": {
          "type": "string",
        },
        "employees": {
          "type": "number",
        },
      },
      "required": [
        "companyName",
        "employees",
      ],
      "type": "object",
    },
    "FlexibleEntity": {
      "properties": {
        "entityUnion": {
          "oneOf": [
            {
              "$ref": "#/definitions/Person",
            },
            {
              "$ref": "#/definitions/Company",
            },
          ],
        },
        "id": {
          "type": "string",
        },
        "mixedUnion": {
          "oneOf": [
            {
              "type": "string",
            },
            {
              "items": {
                "$ref": "#/definitions/Tag",
              },
              "type": "array",
            },
          ],
        },
        "primitiveUnion": {
          "oneOf": [
            {
              "type": "string",
            },
            {
              "type": "number",
            },
            {
              "type": "boolean",
            },
          ],
        },
      },
      "required": [
        "id",
        "primitiveUnion",
        "entityUnion",
      ],
      "type": "object",
    },
    "Person": {
      "properties": {
        "age": {
          "type": "number",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "name",
        "age",
      ],
      "type": "object",
    },
    "Tag": {
      "properties": {
        "color": {
          "type": "string",
        },
        "label": {
          "type": "string",
        },
      },
      "required": [
        "label",
      ],
      "type": "object",
    },
  },
}
`;
